### PR TITLE
Use the owner module instead of origin with UnboundMethod#{inspect,==} 

### PR DESCRIPTION
**Pull request by rwstauner**
We're only concerned with where the method was declared
(rather than the receiver of the instance_method call).

See https://github.com/ruby/ruby/pull/6855
and https://bugs.ruby-lang.org/issues/18798

### CODE REVIEW

File: src/main/ruby/truffleruby/core/unbound_method.rb [hunk 1]
```diff
@@ -10,7 +10,8 @@
 
 class UnboundMethod
   def inspect
-    Truffle::MethodOperations.inspect_method(self, Primitive.unbound_method_origin(self), owner)
+    declaring_module = Primitive.unbound_method_declaring_module(self)
+    Truffle::MethodOperations.inspect_method(self, declaring_module, declaring_module)
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/unbound_method.rb:14]: Could you use `UnboundMethod#owner` here? And then we don't need `unbound_method_declaring_module` at all

Comment (eregon) [reply src/main/ruby/truffleruby/core/unbound_method.rb:14]: Looking at:
```
irb(main):001:0> module A; end
=> nil
irb(main):002:0> module B; end
=> nil
irb(main):003:0> module A; def foo; end; end
=> :foo
irb(main):004:0> A.instance_method(:foo).owner
=> A
irb(main):005:0> B.define_method(:foo, A.instance_method(:foo))
=> :foo
irb(main):006:0> B.instance_method(:foo).owner
=> B
```
It even needs to be the `#owner` to be correct (I think the declaring module would be A for that last line).

Comment (rwstauner) [reply src/main/ruby/truffleruby/core/unbound_method.rb:14]: Interesting, thanks!  The only situation where I've found `owner` to be different from `declaring module` is with this:
```ruby
class A
  private
  def foo; end
end
class B < A
  public :foo
end
puts B.instance_method(:foo).owner
```
and in this case, using `owner` is not the same behavior as CRuby.
I added some specs to confirm that `declaringModule` is required here to match the behavior in CRuby.

Comment (rwstauner) [reply src/main/ruby/truffleruby/core/unbound_method.rb:14]: Which is too bad; it would have been nice to be able to delete the primitive.

Comment (eregon) [reply src/main/ruby/truffleruby/core/unbound_method.rb:14]: Mmh, but that's a case of a "zsuper method", a weird edge case in CRuby we do not try to replicate in TruffleRuby because it seems unnecessarily complex & inconsistent for semantics and AFAIK nobody uses the different behavior.
In such a case TruffleRuby does define a full method on B which is independent from A, so the correct output is `B#foo` (when removing `.owner` in that snippet), even though indeed it differs from CRuby which defines that `B#foo` as `def foo(...) = super(...)` basically but with more edge cases and pretending it half exists.

With:
```ruby
class A
  private
  def foo = 42
end
class B < A
  public :foo
end

p B.instance_method(:foo).owner
p B.instance_method(:foo)

p B.new.foo

A.remove_method :foo

p B.instance_method(:foo)
p B.new.foo
```
TruffleRuby:
```
B
#<UnboundMethod: B#foo() zsuper_def.rb:3>
42
#<UnboundMethod: B#foo() zsuper_def.rb:3>
42
```
CRuby:
```
B
#<UnboundMethod: A#foo() zsuper_def.rb:3>
42
zsuper_def.rb:16:in `instance_method': undefined method `foo' for class `B' (NameError)
zsuper_def.rb:17:in `<main>': undefined method `foo' for #<B:0x00007f2c9b8ebd08> (NoMethodError)
```
I consider CRuby incorrect here, it breaks when there is no need, pretends there is no B#foo when there is and the inconsistency between owner and inspect output which makes no sense, etc.

TLDR let's use `owner`, `UnboundMethod#inspect` and `UnboundMethod#owner` should be consistent.

Comment (eregon) [reply src/main/ruby/truffleruby/core/unbound_method.rb:14]: FWIW there is a long discussion about this inconsistency in CRuby in https://bugs.ruby-lang.org/issues/18729

Comment (rwstauner) [reply src/main/ruby/truffleruby/core/unbound_method.rb:14]: Thanks for sharing that.
I see at the end of that thread is a PR where you added an MRI test that still fails on TruffleRuby, which I was digging into last week.
Ok, so if we ignore the zsuper part then I think owner and declaring module are the same in which case we can simplify this PR down to just owner and leave the zsuper stuff for another day. 👍

**Comment by eregon** [comment]
As a note, we still need the `origin` field because `UnboundMethod#super_method` needs it, this is similar in CRuby.

File: src/main/ruby/truffleruby/core/unbound_method.rb (hunk 1 above)
Comment (eregon) [reply src/main/ruby/truffleruby/core/unbound_method.rb]: Thanks for the changes.
FWIW the (current) plan is to never implement zsuper methods in TruffleRuby, I think it's something that breaks the whole model of a Method/UnboundMethod being a handle to a specific method definition.
I tried very hard to remove it from CRuby but I guess matz likes this weird corner case or worries about compatibility, even though I think basically no code uses the fact the super method can be replaced and the subclass method will then change behavior. It's also inefficient as it causes an extra call (2 vs 1 when calling a zsuper method) and needs an extra cache entry for method lookup.
